xarray/tests/test_backends.py

```diff
@@ -1767,6 +1771,77 @@ def test_setncattr_string(self) -> None:
                 assert_array_equal(one_element_list_of_strings, totest.attrs["bar"])
                 assert one_string == totest.attrs["baz"]
 
+    @pytest.mark.parametrize(
+        "compression",
+        [
+            None,
+            "zlib",
+            "szip",
+            "zstd",
+            "blosc_lz",
+            "blosc_lz4",
+            "blosc_lz4hc",
+            "blosc_zlib",
+            "blosc_zstd",
+        ],
+    )
+    def test_compression_encoding(self, compression: str | None) -> None:
+        if self.nc4_version < Version("1.6.2"):
+            pytest.skip(
+                "Compression options only available for netcdf4-python >= 1.6.2, running with {self.nc_version}."
+            )
+        data = create_test_data(dim_sizes=(20, 40, 10))
+        encoding_params = dict(compression=compression, blosc_shuffle=1)
+        data["var2"].encoding.update(encoding_params)
+        data["var2"].encoding.update(
+            {
+                "chunksizes": (20, 20),
+                "original_shape": data.var2.shape,
+                "blosc_shuffle": 1,
+                "fletcher32": False,
+            }
+        )
+        with self.roundtrip(data) as actual:
+            expected_encoding = data["var2"].encoding.copy()
+            # compression does not appear in the retrieved encoding, that differs
+            # from the input encoding. shuffle also chantges. Here we modify the
+            # expected encoding to account for this
+            compression = expected_encoding.pop("compression")
+            blosc_shuffle = expected_encoding.pop("blosc_shuffle")
+            if compression is not None:
+                if "blosc" in compression and blosc_shuffle:
+                    expected_encoding["blosc"] = {
+                        "compressor": compression,
+                        "shuffle": blosc_shuffle,
+                    }
+                    expected_encoding["shuffle"] = False
+                elif compression == "szip":
+                    expected_encoding["szip"] = {
+                        "coding": "nn",
+                        "pixels_per_block": 8,
+                    }
+                    expected_encoding["shuffle"] = False
+                else:
+                    # This will set a key like zlib=true which is what appears in
+                    # the encoding when we read it.
+                    expected_encoding[compression] = True
+                    if compression == "zstd":
+                        expected_encoding["shuffle"] = False
+            else:
+                expected_encoding["shuffle"] = False
+
+            actual_encoding = actual["var2"].encoding
+            for k, v in expected_encoding.items():
+                assert v == actual_encoding[k]
+        if (
+            encoding_params["compression"] is not None
+            and "blosc" not in encoding_params["compression"]
```

Comment:
My typing-fu isn't competible enough, any hints appreciated:

```
xarray/tests/test_backends.py: note: In member "test_compression_encoding" of class "TestNetCDF4Data":
xarray/tests/test_backends.py:1838: error: Unsupported right operand type for in ("Union[int, str]")  [operator]
```

Reply:
I assume that is because Mypy inferes the the type where you create `encoding_params` as `dict[str, str | int | None]`. And now you are using the `in` operator on something that is potentially an int.
Simply annotate the dict like this:
`encoding_params: dict[str, Any] = dict(...)`